File: app/src/main/java/me/shkschneider/skeleton/ApplicationHelper.java
package me.shkschneider.skeleton;

import android.app.Activity;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.Signature;
import android.content.res.AssetManager;
import android.content.res.Resources;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.provider.Settings;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import me.shkschneider.app.BuildConfig;
import me.shkschneider.app.MainApplication;

public class ApplicationHelper {

    public static boolean debug() {
        return BuildConfig.DEBUG;
    }

    public static Resources resources() {
        return MainApplication.CONTEXT.getResources();
    }

    public static AssetManager assets() {
        return MainApplication.CONTEXT.getAssets();
    }

    public static String packageName() {
        return MainApplication.CONTEXT.getPackageName();
    }

    public static String name() {
        try {
            final PackageManager packageManager = MainApplication.CONTEXT.getPackageManager();
            if (packageManager == null) {
                LogHelper.warning("PackageManager was NULL");
                return null;
            }

            final ApplicationInfo applicationInfo = packageManager.getApplicationInfo(packageName(), 0);
            if (applicationInfo == null) {
                LogHelper.warning("ApplicationInfo was NULL");
                return null;
            }

            final CharSequence label = applicationInfo.loadLabel(packageManager);
            if (label == null) {
                LogHelper.warning("Label was NULL");
                return null;
            }

            return label.toString();
        }
        catch (final Exception e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    public static String versionName() {
        try {
            final PackageManager packageManager = MainApplication.CONTEXT.getPackageManager();
            if (packageManager == null) {
                LogHelper.warning("PackageManager was NULL");
                return null;
            }

            return packageManager.getPackageInfo(packageName(), PackageManager.GET_META_DATA).versionName;
        }
        catch (final Exception e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    public static int versionCode() {
        try {
            final PackageManager packageManager = MainApplication.CONTEXT.getPackageManager();
            if (packageManager == null) {
                LogHelper.warning("PackageManager was NULL");
                return -1;
            }

            return packageManager.getPackageInfo(packageName(), PackageManager.GET_META_DATA).versionCode;
        }
        catch (final Exception e) {
            LogHelper.wtf(e);
            return -1;
        }
    }

    public static Drawable icon() {
        try {
            final PackageManager packageManager = MainApplication.CONTEXT.getPackageManager();
            if (packageManager == null) {
                LogHelper.warning("PackageManager was NULL");
                return null;
            }

            final ApplicationInfo applicationInfo = packageManager.getApplicationInfo(packageName(), 0);
            if (applicationInfo == null) {
                LogHelper.warning("ApplicationInfo was NULL");
                return null;
            }

            return applicationInfo.loadIcon(packageManager);
        }
        catch (final Exception e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    public static List<String> permissions() {
        try {
            final PackageManager packageManager = MainApplication.CONTEXT.getPackageManager();
            if (packageManager == null) {
                LogHelper.warning("PackageManager was NULL");
                return null;
            }

            final PackageInfo packageInfo = packageManager.getPackageInfo(ApplicationHelper.packageName(), PackageManager.GET_PERMISSIONS);

            final List<String> list = new ArrayList<String>();
            Collections.addAll(list, packageInfo.requestedPermissions);
            return list;
        }
        catch (final Exception e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    public static String signature() {
        final PackageManager packageManager = MainApplication.CONTEXT.getPackageManager();
        if (packageManager == null) {
            LogHelper.warning("PackageManager was NULL");
            return null;
        }

        try {
            final PackageInfo packageInfo = packageManager.getPackageInfo(MainApplication.CONTEXT.getPackageName(), PackageManager.GET_SIGNATURES);
            if (packageInfo == null) {
                LogHelper.warning("PackageInfo was NULL");
                return null;
            }

            final Signature[] signatures = packageInfo.signatures;
            if (signatures == null) {
                LogHelper.debug("No signatures");
                return null;
            }

            return signatures[0].toCharsString();
        }
        catch (final Exception e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    public static boolean settings(final Activity activity) {
        final int api = AndroidHelper.api();
        final Intent intent = new Intent();
        intent.setAction(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
        intent.setData(Uri.fromParts("package", packageName(), null));
        activity.startActivity(intent);
        return true;
    }

}

File: android/src/main/kotlin/me/shkschneider/skeleton/android/app/ApplicationHelper.kt
package me.shkschneider.skeleton.android.app

import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.content.pm.ApplicationInfo
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.content.res.Resources
import android.graphics.drawable.Drawable
import android.os.Build
import me.shkschneider.skeleton.android.log.Logger
import me.shkschneider.skeleton.android.os.AndroidHelper
import me.shkschneider.skeleton.android.provider.SystemServices
import me.shkschneider.skeleton.android.provider.ContextProvider
import me.shkschneider.skeleton.kotlin.jvm.tryOr
import me.shkschneider.skeleton.kotlin.jvm.tryOrNull

object ApplicationHelper {

    // ApplicationInfo.loadDefaultIcon(packageManager)
    const val DEFAULT_ICON = android.R.drawable.sym_def_app_icon
    const val INSTALLER = "com.android.vending"

    // <https://stackoverflow.com/a/25517680/603270>
    val debuggable: Boolean
        get() = SkeletonApplication.DEBUGGABLE

    // Resources.getSystem()
    val resources: Resources
        get() = ContextProvider.applicationContext().resources

    val packageName: String
        get() = ContextProvider.applicationContext().packageName

    val packageManager: PackageManager
        get() = ContextProvider.applicationContext().packageManager

    fun applicationInfo(packageName: String = ApplicationHelper.packageName): ApplicationInfo? =
        tryOrNull {
            packageManager.getApplicationInfo(packageName, 0)
        }

    fun name(packageName: String = ApplicationHelper.packageName): String? =
        applicationInfo(packageName)?.loadLabel(packageManager)?.toString() ?: run {
            Logger.warning("Label was NULL")
            null
        }

    fun exists(packageName: String = ApplicationHelper.packageName): Boolean =
        tryOr({
            packageManager.getApplicationInfo(packageName, 0)
            true
        }, false) == true

    fun packageInfo(packageName: String = ApplicationHelper.packageName, flags: Int = 0): PackageInfo? =
        tryOrNull {
            packageManager.getPackageInfo(packageName, flags)
        }

    fun versionName(packageName: String = ApplicationHelper.packageName): String? =
        packageInfo(packageName, PackageManager.GET_META_DATA)?.versionName ?: run {
            Logger.warning("VersionName was NULL")
            return null
        }

    @SuppressLint("NewApi")
    fun versionCode(packageName: String = ApplicationHelper.packageName): Long? {
        val packageInfo = packageInfo(packageName, PackageManager.GET_META_DATA)
        if (Build.VERSION.SDK_INT >= 28) {
            return packageInfo?.longVersionCode ?: run {
                Logger.warning("LongVersionCode was NULL")
                return null
            }
        } else {
            @Suppress("DEPRECATION")
            return packageInfo?.versionCode?.toLong() ?: run {
                Logger.warning("VersionCode was NULL")
                return null
            }
        }
    }

    fun buildTime(packageName: String = ApplicationHelper.packageName): Long? =
        packageInfo(packageName, PackageManager.GET_META_DATA)?.lastUpdateTime ?: run {
            Logger.warning("LastUpdateTime was NULL")
            return null
        }

    fun permissions(packageName: String = ApplicationHelper.packageName): List<String>? =
        packageInfo(packageName, PackageManager.GET_PERMISSIONS)?.requestedPermissions?.toList() ?: run {
            Logger.warning("RequestedPermissions was NULL")
            return null
        }

    fun icon(packageName: String = ApplicationHelper.packageName): Int? =
        tryOrNull {
            val applicationInfo = packageManager.getApplicationInfo(packageName, 0)
            applicationInfo.icon
        }

    fun drawable(packageName: String = ApplicationHelper.packageName) : Drawable? =
        tryOrNull {
            val applicationInfo = packageManager.getApplicationInfo(packageName, 0)
            applicationInfo.loadIcon(packageManager)
        }

    @TargetApi(AndroidHelper.API_19)
    fun clear() {
        Logger.info("clearApplicationUserData()")
        SystemServices.activityManager()?.clearApplicationUserData()
    }

}
